### PR TITLE
CT-2234 Offender SARs : create steps

### PR DESCRIPTION
**Pull request by Polsonby**
This is a proposal for how to manage steps-based creation of a new Offender SAR

I've tried to keep as much as possible consistent with existing structure and approach. The new bit is a separate controller inheriting from CasesController and a route that catches "osar" cases creation and supports steps. Steps are defined within the OffenderSAR model and supported by a Steppable module which adds a writer and a few simple methods for working out the next step/previous step. 

Go to /cases/new and you'll see a new hard-coded entry there (pending the creation of a seed user/team for Branston that we can log in with) for Offender SARs. 
image

You can view the first, second, third, and fourth pages of the sequence. It presently loops around, but it won't when we're done with validating and creating new Offender SAR Cases.

http://localhost:3000/cases/new/osar
http://localhost:3000/cases/new/osar/requester-details
http://localhost:3000/cases/new/osar/requested-info
http://localhost:3000/cases/new/osar/date-received

image
image
image
image

This uses the create action of the new offender_sars_controller to "ratchet" through the steps for creating the new model each time it's submitted (hard coded to assume successful validation presently) 

I've also updated the readme file with a couple new things I found out!

https://github.com/ministryofjustice/correspondence_tool_staff/tree/CT-2234-offender-sars-create-steps
- new sections about data migrations, Localisation keys checking, and the Site prism page manifest file

### CODE REVIEW

**Review by mseedat-moj — CHANGES_REQUESTED** [review]
Good start. Discussion with @Polsonby:

1. Replace `osar` with `OffenderSAR` or similar
2. Use the `FeatureSet` feature flag for `offender_sars` as found in `Settings.yml`

File: app/views/admin/cases/select_type.html.slim [hunk 1]
```diff
@@ -11,3 +11,6 @@ ul#create-correspondences
     li style="margin-bottom: 15px"
       = link_to correspondence_type.abbreviation_and_name,
                 new_admin_case_path(correspondence_type: correspondence_type.abbreviation.downcase)
+  li style="margin-bottom: 15px"
+  	= link_to "OFFENDER_SAR - test link",
+                new_case_path("offender_sar")
```
Comment (Polsonby) [inline app/views/admin/cases/select_type.html.slim]: oops - changes to this file are redundant, need to remove

File: app/controllers/cases/offender_sars_controller.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    elsif params[:commit]
+      obj.next_step
+    end
+  end
+
+  def get_step_partial(current_step)
+    step_name = current_step.split("/").first.tr('-', '_')
+    "#{step_name}_step"
+  end
+end
```
Comment (Polsonby) [inline app/controllers/cases/offender_sars_controller.rb]: This get_step_partial is redundant, needs to be removed

**Comment by Polsonby** [comment]
Hey, I've made those changes, please have another look when you have a chance :) 

I spotted a couple of things got left in by mistake, sorry - have commented and I'll remove tomorrow along with the conflict in rails_helper

File: spec/features/cases/osar/case_creating_spec.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
+require 'rails_helper'
```
Comment (Polsonby) [inline spec/features/cases/osar/case_creating_spec.rb]: the location of this file needs to be updated

File: spec/site_prism/page_objects/pages/cases/new/osar_page_date_received.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
+module PageObjects
```
Comment (Polsonby) [inline spec/site_prism/page_objects/pages/cases/new/osar_page_date_received.rb]: file needs to be renamed

File: spec/site_prism/page_objects/pages/cases/new/osar_page_requested_info.rb [hunk 5]
```diff
@@ -0,0 +1,28 @@
+module PageObjects
```
Comment (Polsonby) [inline spec/site_prism/page_objects/pages/cases/new/osar_page_requested_info.rb]: file name

File: spec/site_prism/page_objects/pages/cases/new/osar_page_requester_details.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+module PageObjects
```
Comment (Polsonby) [inline spec/site_prism/page_objects/pages/cases/new/osar_page_requester_details.rb]: file name

File: spec/site_prism/page_objects/pages/cases/new/osar_page_subject_details.rb [hunk 7]
```diff
@@ -0,0 +1,28 @@
+module PageObjects
```
Comment (Polsonby) [inline spec/site_prism/page_objects/pages/cases/new/osar_page_subject_details.rb]: file name

**Review by Polsonby — COMMENTED** [review]
need to rename a few files still

**Pull request closed without merging**